src/clj/puppetlabs/services/request_handler/request_handler_core.clj

```diff
@@ -12,6 +12,14 @@
             [slingshot.slingshot :as sling]))
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Constants
+
+(def error-message-404
+  (str "Error: This master is running Puppet 4, and the URL format has changed. "
+       "This request was likely made with a Puppet 3 Agent, as it used an invalid "
+       "URL format. Please upgrade your agents to Puppet 4 and try again."))
```

Comment:
Handling "known" URLs would involve all of the same problems that we're trying to get away from, about having to camp on the entire "/" URL namespace... because the first segment of the URL from a puppet 3 agent is going to be the environment name, which could be literally anything.

So, the approach we agreed on as 'best effort' for now is to just add a catch-all 404 handler that has an error message that is suitable for cases where the issue *is* related to a Puppet 3 agent, but also to cases where the issue is *not* related to a Puppet 3 agent.  So we may want to soffent up some of the wording, e.g. change "likely" to "possibly", add in something like "If your agent is puppet 3, try upgrading and running the agent again.", etc.